Blueprints for template entities

tetele opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->
No

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add blueprint capabilities for `template` entities, as per https://github.com/home-assistant/architecture/discussions/1027

```yaml
#/config/blueprints/template/simple_invert.yaml
blueprint:
  name: Invert a binary sensor
  description: Creates a binary_sensor which holds the inverted value of a reference binary_sensor
  domain: template
  source_url: http://example.com # used for updates
  input:
    original_entity:
      name: Binary sensor to be inverted
      description: The binary_sensor which needs to have its value inverted
      selector:
        entity:
          domain: binary_sensor
variables:
  original_entity: !input original_entity
binary_sensor:
  state: >
    {% if states(original_entity) == 'on' %}
      off
    {% elif states(original_entity) == 'off' %}
      on
    {% else %}
      {{ states(original_entity) }}
    {% endif %}
  # delay_on: not_used in this example
  # delay_off: not_used in this example
  # auto_off: not_used in this example
  availability: "{{ states(original_entity) not in ('unknown', 'unavailable') }}"
```
```yaml
#/config/configuration.yaml
...
template:
  - use_blueprint:
      path: simple_invert.yaml
      input:
        original_entity: binary_sensor.foo
    name: Inverted foo
    unique_id: inverted_binary_sensor_foo
  - use_blueprint:
      path: simple_invert.yaml
      input:
        original_entity: binary_sensor.bar
    name: Inverted bar
    unique_id: inverted_binary_sensor_bar
```

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: https://github.com/home-assistant/architecture/discussions/1027
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/35057

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

tetele committed:
Template domain blueprints

tetele committed:
Default blueprint for templates

tetele committed:
Some linting

tetele committed:
Template entity updates

tetele committed:
Load and use blueprints in config

tetele committed:
Added missing mapping methods for templates

tetele committed:
Linting

tetele committed:
Added tests

tetele committed:
Wrong schema type

tetele committed:
Hassfest errors

tetele committed:
More linting issues

tetele committed:
Merge branch 'dev' into template-blueprint

tetele committed:
Refactor based on desired schema

In the [architecture discussion](https://github.com/home-assistant/architecture/discussions/1027), the template blueprint instance did not specify the platform (e.g. `binary_sensor`), but the initial implementation assumed that schema.

tetele committed:
Merge branch 'dev' into template-blueprint

tetele committed:
Create default template blueprints on first run

tetele committed:
Moved TemplateConfig definition

This is to avoid circular references

tetele committed:
Corrected methods to find templates based on blueprints

tetele committed:
Corrected missing entity config information

tetele committed:
Added tests

MartinHjelmare commented on homeassistant/components/template/__init__.py:
```diff
@@ -29,13 +30,61 @@
+        template_entity.entity_id
+        for platform in PLATFORMS
+        if platform in hass.data
+        for template_entity in hass.data[platform].entities
```
We're not allowed to access another integration's `hass.data` items directly. We can use this helper to get all platforms of the template integration.

https://github.com/home-assistant/core/blob/9921a67a05bdb1b1a17b08b1bebbde7f90b250c3/homeassistant/helpers/entity_platform.py#L1063-L1066

The `EntityPlatform` has a dict with all its entities in the `entities` attribute.

tetele committed:
Don't use hass.data

Address comments https://github.com/home-assistant/core/pull/126971/#discussion_r1780097187

MartinHjelmare commented on homeassistant/components/template/__init__.py:
```diff
@@ -29,13 +31,52 @@
+    """Return the blueprint the template is based on or None."""
+    for platform in async_get_platforms(hass, DOMAIN):
+        if template_entity := platform.entities.get(entity_id):
+            if isinstance(template_entity, TemplateEntity):
```
Combine these two conditions with `and`.

tetele replied:
I can't use `and` since apparently the walrus does not make the assigned variable available right away in the same expression, but I have combined them in a single `if`

MartinHjelmare replied:
It works by using a parenthesis. But the current way works too.

tetele committed:
Prevent creating blueprints during testing

tetele committed:
Combine 2 ifs

Address comment https://github.com/home-assistant/core/pull/126971/#discussion_r1780160870

tetele committed:
Improve test coverage

tetele commented:
I don't understand why [this](https://github.com/home-assistant/core/actions/runs/11096197325/job/30825824664?pr=126971) keeps happening, as I've explicitly added [a fixture](https://github.com/home-assistant/core/pull/126971/files#diff-6fa7c7f977eb9a7b189941a3749e9113714383f71605839d9394201651824dcfR41-R43) for the purpose of keeping the tests clean and locally the blueprint file does not get copied.

tetele committed:
Prevent template component from dirtying test env

tetele commented on homeassistant/components/template/config.py:
```diff
@@ -39,6 +58,9 @@
         vol.Optional(CONF_CONDITION): cv.CONDITIONS_SCHEMA,
         vol.Optional(CONF_ACTION): cv.SCRIPT_SCHEMA,
+        vol.Optional(CONF_VARIABLES): cv.SCRIPT_VARIABLES_SCHEMA,
+        vol.Optional("raw_blueprint_inputs"): cv.match_all,
+        vol.Optional("raw_config"): cv.match_all,
```
Actually, I've refactored so that these are no longer needed, but forgot to remove them. I'll do so in a following PR.

tetele committed:
Remove useless hard-coded validation

tetele committed:
Merge branch 'dev' into template-blueprint

tetele commented on homeassistant/components/template/helpers.py:
```diff
@@ -0,0 +1,40 @@
+
+async def _reload_blueprint_templates(hass: HomeAssistant, blueprint_path: str) -> None:
+    """Reload all templates that rely on a specific blueprint."""
+    await hass.services.async_call(DOMAIN, SERVICE_RELOAD)
```
This line is not covered with tests as the RELOAD handler is [locally defined](https://github.com/home-assistant/core/blob/cf7205584e67c05dd04e3d718e45b5be8d27b68a/homeassistant/components/template/__init__.py#L84) in `async_setup`

Should I take it out as a separate method and patch it to test this as well?

MartinHjelmare replied:
Since the reload helper should be tested separately, I think it's enough to test that this callback calls the service. We can do that by mocking the service with that test helper, using the same service domain and name.

https://github.com/home-assistant/core/blob/781c3eed2fa95ee8f765761e9b40f578df4cca2a/tests/common.py#L346C5-L355

tetele committed:
Improve code coverage to 100%

tetele commented on homeassistant/components/template/config.py:
```diff
@@ -39,6 +58,7 @@
         vol.Optional(CONF_TRIGGER): cv.TRIGGER_SCHEMA,
```
As it stands, you can't use Jinja templates in triggers. It would be great to be able to do so, but i think that would mean moving the logic for registering triggers much later. Is that OK?

Here's an example:
```yaml
blueprint:
  domain: template
  input:
    tracked_entity:
      name: Tracked entity
      description: The `sensor` which needs to change from `something` to `something_else` for the resulting `binary_sensor` to turn `on`
      selector:
        entity:
variables:
  tracked_entity: !input tracked_entity
trigger:
  - platform: state
    entity_id: "{{ tracked_entity }}" # error: `Invalid config for 'template' at configuration.yaml, line 88: Entity {{ tracked_entity }} is neither a valid entity ID nor a valid UUID for dictionary value 'entity_id', got None`
    from: "something"
    to: "something_else"
binary_sensor:
  ...
```

`trigger_variables` might be needed, but the question is when should they be evaluated? Is it safe to do so at config validation? Or should the trigger registration be postponed to much later, when it's safe to evaluate `trigger_variables` as regular tempaltes?

MartinHjelmare replied:
This isn't specific to blueprints, right? If it's not specific to blueprints I suggest discussing it separately in a new architecture discussion.

tetele replied:
I'd say that template blueprints expose the missing feature. In automations, you can use `trigger_variables` specifically for allowing the templating of `entity_id` in blueprints, which had not existed for templates until now.

> The second variant is setting variables that are available when attaching a trigger when the trigger can contain templated values. These are defined using the trigger_variables key at an automation level. These variables can only contain [limited templates](https://www.home-assistant.io/docs/configuration/templating/#limited-templates). The triggers will not re-apply if the value of the template changes. Trigger variables are a feature meant to support using blueprint inputs in triggers.

([source](https://www.home-assistant.io/docs/automation/trigger/#trigger-variables))

I'm a but stumped about what to do here. If you think this discussion is outside the scope of this feature, I'll start a new architecture discussion, but again... template blueprints _uncovered_ this issue, so i'd say it's related.

MartinHjelmare replied:
I don't think it'd related since it's about triggers and not blueprints.

MartinHjelmare replied:
We don't need to solve all problems at once.

tetele replied:
https://github.com/home-assistant/architecture/discussions/1153

MartinHjelmare commented on homeassistant/components/template/config.py:
```diff
@@ -39,6 +58,7 @@
         vol.Optional(CONF_TRIGGER): cv.TRIGGER_SCHEMA,
         vol.Optional(CONF_CONDITION): cv.CONDITIONS_SCHEMA,
         vol.Optional(CONF_ACTION): cv.SCRIPT_SCHEMA,
+        vol.Optional(CONF_VARIABLES): cv.SCRIPT_VARIABLES_SCHEMA,
```
Don't forget to document this since it will be available for all template entities and not only blueprint templates.

tetele replied:
https://github.com/tetele/home-assistant.io/blob/8f0d137550888bc76fb12d6f28aa6df4a3321e2d/source/_integrations/template.markdown?plain=1#L146-L154

MartinHjelmare commented on homeassistant/components/template/template_entity.py:
```diff
@@ -77,6 +80,7 @@
     {
         vol.Optional(CONF_ATTRIBUTES): vol.Schema({cv.string: cv.template}),
         vol.Optional(CONF_AVAILABILITY): cv.template,
+        vol.Optional(CONF_VARIABLES): cv.SCRIPT_VARIABLES_SCHEMA,
```
Don't forget to document this since it will be available for all template entities and not only blueprint templates.

MartinHjelmare commented on tests/components/template/test_blueprint.py:
```diff
@@ -0,0 +1,235 @@
+    assert foo_template is None
+    assert inverted_foo_template == "inverted_binary_sensor.yaml"
+
+    inverted_binary_sensor_blueprint_instances = template.templates_with_blueprint(
```
```suggestion
    inverted_binary_sensor_blueprint_entity_ids = template.templates_with_blueprint(
```

MartinHjelmare commented on tests/components/template/test_blueprint.py:
```diff
@@ -0,0 +1,235 @@
+    inverted_binary_sensor_blueprint_instances = template.templates_with_blueprint(
+        hass, "inverted_binary_sensor.yaml"
+    )
+    assert len(inverted_binary_sensor_blueprint_instances) == 2
```
```suggestion
    assert len(inverted_binary_sensor_blueprint_entity_ids) == 2
```

MartinHjelmare commented on tests/components/template/test_blueprint.py:
```diff
@@ -0,0 +1,235 @@
+            },
+        )
+
+    assert "more than one platform defined per blueprint" in caplog.text
```
This error log is confusing since there's no platform at all defined, right?

tetele replied:
That above is the blueprint instance. The blueprint has 2 platforms defined https://github.com/tetele/ha-core/blob/2d678a9552908cdf7863d3af9e475ce4161c8d22/tests/components/template/test_blueprint.py#L68-L69

MartinHjelmare replied:
Ok

tetele commented on homeassistant/components/template/helpers.py:
```diff
@@ -0,0 +1,40 @@
+
+def _blueprint_in_use(hass: HomeAssistant, blueprint_path: str) -> bool:
+    """Return True if any template references the blueprint."""
+    from . import templates_with_blueprint  # pylint: disable=import-outside-toplevel
```
Same for `blueprint_in_template`, i suppose?

tetele commented on tests/components/template/test_blueprint.py:
```diff
@@ -0,0 +1,235 @@
+            {
+                "template": [
+                    {"binary_sensor": {"name": "test entity", "state": "off"}},
+                    # A real sensor is needed
```
I can't remember 😅 Must be a leftover, I'll take it out

tetele committed:
Address review comments

tetele committed:
Moved helpers in helpers.py

As per comment https://github.com/home-assistant/core/pull/126971#discussion_r1786539889

tetele commented on homeassistant/components/template/blueprints/inverted_binary_sensor.yaml:
```diff
@@ -0,0 +1,27 @@
+  name: Invert a binary sensor
+  description: Creates a binary_sensor which holds the inverted value of a reference binary_sensor
+  domain: template
+  source_url: http://example.com
```
This needs to be updated before merging

tetele committed:
Fix blueprint source URL

MartinHjelmare committed:
Merge branch 'dev' into template-blueprint

MartinHjelmare reviewed (approved):
Great work!

tetele commented:
Thank you very much!
